### PR TITLE
Save enum values as string in settings

### PR DESCRIPTION
**Pull request by 3nids**
This allows to save enum values as strings in the settings, making it clearer when looking into advanced panel.
There is no API break but this might cause issue in Python since one would need to use `enumValue` to get the enum value as `value` would now return a string and not an integer anymore.
From what I see, I think the main issue would be the snapping settings which might be used by plugin devs.

Python API still has to be made, just waiting for feedback if this is reasonable or not.

If not, we could introduce the API (for new enums) and do the switch in next major version.

### CODE REVIEW

**Comment by nyalldawson** [comment]
Nice!

The more I think about it, the more I think we need some way to "register" settings. Then we could provide a little descriptive text for them (shown in tooltips in the advanced settings editor) AND we could build off this PR to show combo-boxes for enum-type settings too (by registering the enum associated with the setting). We could then also generate the whole global settings ini through a script too...

File: src/core/qgssettings.h [hunk 1]
```diff
@@ -224,33 +225,126 @@ class CORE_EXPORT QgsSettings : public QObject
...
      */
     template <class T>
     T enumValue( const QString &key, const T &defaultValue,
-                 const Section section = NoSection, bool flag = false ) const
+                 const Section section = NoSection ) const
     {
+      QMetaEnum metaEnum = QMetaEnum::fromType<T>();
+      if ( !metaEnum.isValid() )
+      {
+        QgsDebugMsg( "Invalid metaenum. Enum probably misses Q_ENUM or Q_FLAG declaration." );
```
Comment (m-kuhn) [inline src/core/qgssettings.h]: Could this be a `Q_ASSERT` to prevent devs from using this the wrong way?

Comment (m-kuhn) [reply src/core/qgssettings.h]: or is the usage widespread throughout the code and we need this for legacy fallback?

File: src/core/qgssettings.h [hunk 2]
```diff
@@ -224,33 +225,126 @@ class CORE_EXPORT QgsSettings : public QObject
...
+                       const Section section = NoSection )
+    {
+      QMetaEnum metaEnum = QMetaEnum::fromType<T>();
+      if ( metaEnum.isValid() )
+      {
+        setValue( key, metaEnum.valueToKeys( value ), section );
+      }
+      else
+      {
+        QgsDebugMsg( "Invalid metaenum. Enum probably misses Q_ENUM or Q_FLAG declaration." );
```
Comment (m-kuhn) [inline src/core/qgssettings.h]: Might be good to include `key` and `value` in the debug message.

**Comment by m-kuhn** [comment]
Good job!

> The more I think about it, the more I think we need some way to "register" settings

I agree an optional registration for settings would be great.

**Pull request merged**